src/workspace.ts

```diff
@@ -154,36 +163,34 @@ export class Workspace {
       fs.mkdirSync(workspaceDir);
     }
 
-    // If a folder exists for a workspace repo and it can't be opened with
-    // nodegit, we need to remove it.  This happens when there's not a --fresh
+    // If a folder exists for a workspace repo and it can't be opened,
+    // we need to remove it.  This happens when there's not a --fresh
     // invocation and bower installed the dependency instead of git.
-    for (const repo of repos) {
+    await Promise.all(repos.map(async (repo) => {
       if (existsSync(repo.dir) && !repo.git.isGit()) {
         if (options.verbose) {
           console.log(`Removing existing folder: ${repo.dir}...`);
         }
         await rimraf(repo.dir);
       }
-    }
+    }));
   }
 
   /**
    * Given all the repos defined in the workspace, lets iterate through them
    * and either clone them or update their clones and set them to the specific
    * refs.
-   * TODO(fks) 09-25-2017: Better error handling. Standardize/format errors
-   * so that single type of error thrown.
    */
   private async _cloneOrUpdateWorkspaceRepos(repos: WorkspaceRepo[]) {
-    for (const repo of repos) {
+    return batchProcess(repos, async (repo: WorkspaceRepo) => {
```

Comment:
no action needed: I wonder if what we really want to rate limit here is outstanding fine-grained tasks, not higher-level tasks. ie, in `GitRepo`, automatically rate limit:

```ts
  async fetch(remoteName?: string): Promise<ExecResult> {
    // ...
    return gitLimiter.schedule(() => exec(this.dir, 'git', commandArgs));
  }
```

Then the higher-level tasks will be naturally rate-limited by the limits on their underlying operations.

Reply:
That's interesting, much closer to the actual interface of the bottleneck library. 

It doesn't really solves rate limiting though, you could have each of the 3 rate limiters going off at once and slowing each other. Also harder to coordinate with raw calls to `exec()`.